File: app/src/main/java/com/byd/dashcast/infrastructure/task/ProxyTaskFinder.java
package com.byd.dashcast.infrastructure.task;

import com.byd.dashcast.AppLogger;
import com.byd.dashcast.proxy.ProxyClient;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Paths 2 + 2b: daemon-side {@code dumpsys activity recents} and
 * {@code dumpsys activity activities} via the shell-uid proxy.
 *
 * Requires the proxy daemon to be connected ({@link ProxyClient#isConnected()}).
 * Uses the same regex logic migrated verbatim from {@code ClusterService} to
 * preserve identical parse behaviour across all BYD ROM variants.
 */
public final class ProxyTaskFinder implements TaskFinder {

    private static final String TAG = "ProxyTaskFinder";

    // Pre-compiled patterns — same as ClusterService originals (zero logic change).
    private static final Pattern P_TASK_ID =
            Pattern.compile("Task\\{[^}]*#(\\d+)");
    private static final Pattern P_RECENT_SPLIT =
            Pattern.compile("(?m)^\\s*\\* Recent #\\d+:\\s*");

    @Override
    public int findTaskId(String packageName) throws TaskFinderException {
        if (!ProxyClient.isConnected()) {
            throw new TaskFinderException("daemon not connected");
        }
        // Path 2 — dumpsys activity recents
        try {
            String out = ProxyClient.runShell("dumpsys activity recents");
            if (out != null && !out.isEmpty()) {
                int id = parseFromRecents(out, packageName);
                if (id != NOT_FOUND) return id;
                AppLogger.d(TAG, packageName + " not in recents (len=" + out.length() + ")");
            }
        } catch (Throwable t) {
            AppLogger.w(TAG, "recents dump failed: " + t.getMessage());
        }
        // Path 2b — dumpsys activity activities (launcher-agnostic fallback)
        try {
            String out = ProxyClient.runShell("dumpsys activity activities");
            if (out != null && !out.isEmpty()) {
                int id = parseFromActivities(out, packageName);
                if (id != NOT_FOUND) return id;
                AppLogger.d(TAG, packageName + " not in activities (len=" + out.length() + ")");
            }
        } catch (Throwable t) {
            AppLogger.w(TAG, "activities dump failed: " + t.getMessage());
        }
        return NOT_FOUND;
    }

    // ── Parsers (migrated verbatim from ClusterService static methods) ─────────

    public static int parseFromRecents(String dump, String packageName) {
        if (dump == null || packageName == null) return NOT_FOUND;
        // Fast path — affinity on the same Task{...} line.
        try {
            Pattern p = Pattern.compile(
                    "Task\\{[^}]*#(\\d+)[^}]*\\bA=" + Pattern.quote(packageName) + "\\b");
            Matcher m = p.matcher(dump);
            if (m.find()) {
                try { return Integer.parseInt(m.group(1)); } catch (NumberFormatException ignored) {}
            }
        } catch (Exception ignored) {}

        // Block scan — realActivity= or cmp= match anywhere inside the Task block.
        try {
            String[] blocks = P_RECENT_SPLIT.split(dump);
            String marker1 = "realActivity=" + packageName + "/";
            String marker2 = "cmp=" + packageName + "/";
            String marker3 = " A=" + packageName + " ";
            for (String block : blocks) {
                if (block == null || block.isEmpty()) continue;
                if (block.contains(marker1) || block.contains(marker2) || block.contains(marker3)) {
                    Matcher mm = P_TASK_ID.matcher(block);
                    if (mm.find()) {
                        try { return Integer.parseInt(mm.group(1)); } catch (NumberFormatException ignored) {}
                    }
                }
            }
        } catch (Exception ignored) {}
        return NOT_FOUND;
    }

    public static int parseFromActivities(String dump, String packageName) {
        if (dump == null || packageName == null) return NOT_FOUND;
        try {
            Pattern p = Pattern.compile(
                    "ActivityRecord\\{[^}]*\\s"
                            + Pattern.quote(packageName)
                            + "/[^\\s}]+\\st(\\d+)\\}");
            Matcher m = p.matcher(dump);
            if (m.find()) {
                try { return Integer.parseInt(m.group(1)); } catch (NumberFormatException ignored) {}
            }
        } catch (Exception ignored) {}
        return NOT_FOUND;
    }
}

File: app/src/main/java/com/byd/dashcast/infrastructure/task/ProxyTaskFinder.kt
package com.byd.dashcast.infrastructure.task

import com.byd.dashcast.proxy.ProxyClient
import com.byd.dashcast.util.AppLogger

import java.util.regex.Pattern

/**
 * Paths 2 + 2b: daemon-side `dumpsys activity recents` and
 * `dumpsys activity activities` via the shell-uid proxy.
 *
 * Requires the proxy daemon to be connected ([ProxyClient.isConnected]).
 * Uses the same regex logic migrated verbatim from `ClusterService` to
 * preserve identical parse behaviour across all BYD ROM variants.
 */
class ProxyTaskFinder : TaskFinder {

    @Throws(TaskFinder.TaskFinderException::class)
    override fun findTaskId(packageName: String): Int {
        if (!ProxyClient.isConnected()) {
            throw TaskFinder.TaskFinderException("daemon not connected")
        }
        // Path 2 — dumpsys activity recents
        try {
            val out = ProxyClient.runShell("dumpsys activity recents")
            if (!out.isNullOrEmpty()) {
                val id = parseFromRecents(out, packageName)
                if (id != TaskFinder.NOT_FOUND) return id
                AppLogger.d(TAG, "$packageName not in recents (len=${out.length})")
            }
        } catch (t: Throwable) {
            AppLogger.w(TAG, "recents dump failed: ${t.message}")
        }
        // Path 2b — dumpsys activity activities (launcher-agnostic fallback)
        try {
            val out = ProxyClient.runShell("dumpsys activity activities")
            if (!out.isNullOrEmpty()) {
                val id = parseFromActivities(out, packageName)
                if (id != TaskFinder.NOT_FOUND) return id
                AppLogger.d(TAG, "$packageName not in activities (len=${out.length})")
            }
        } catch (t: Throwable) {
            AppLogger.w(TAG, "activities dump failed: ${t.message}")
        }
        return TaskFinder.NOT_FOUND
    }

    companion object {
        private const val TAG = "ProxyTaskFinder"

        // Pre-compiled patterns — same as ClusterService originals (zero logic change).
        private val P_TASK_ID = Pattern.compile("Task\\{[^}]*#(\\d+)")
        private val P_RECENT_SPLIT = Pattern.compile("(?m)^\\s*\\* Recent #\\d+:\\s*")

        // ── Parsers (migrated verbatim from ClusterService static methods) ─────────

        @JvmStatic
        fun parseFromRecents(dump: String?, packageName: String?): Int {
            if (dump == null || packageName == null) return TaskFinder.NOT_FOUND
            // Fast path — affinity on the same Task{...} line.
            try {
                val p = Pattern.compile(
                    "Task\\{[^}]*#(\\d+)[^}]*\\bA=" + Pattern.quote(packageName) + "\\b"
                )
                val m = p.matcher(dump)
                if (m.find()) {
                    try {
                        return m.group(1)!!.toInt()
                    } catch (ignored: NumberFormatException) {
                    }
                }
            } catch (ignored: Exception) {
            }

            // Block scan — realActivity= or cmp= match anywhere inside the Task block.
            try {
                val blocks = P_RECENT_SPLIT.split(dump)
                val marker1 = "realActivity=$packageName/"
                val marker2 = "cmp=$packageName/"
                val marker3 = " A=$packageName "
                for (block in blocks) {
                    if (block.isNullOrEmpty()) continue
                    if (block.contains(marker1) || block.contains(marker2) || block.contains(marker3)) {
                        val mm = P_TASK_ID.matcher(block)
                        if (mm.find()) {
                            try {
                                return mm.group(1)!!.toInt()
                            } catch (ignored: NumberFormatException) {
                            }
                        }
                    }
                }
            } catch (ignored: Exception) {
            }
            return TaskFinder.NOT_FOUND
        }

        @JvmStatic
        fun parseFromActivities(dump: String?, packageName: String?): Int {
            if (dump == null || packageName == null) return TaskFinder.NOT_FOUND
            try {
                val p = Pattern.compile(
                    "ActivityRecord\\{[^}]*\\s" + Pattern.quote(packageName) + "/[^\\s}]+\\st(\\d+)\\}"
                )
                val m = p.matcher(dump)
                if (m.find()) {
                    try {
                        return m.group(1)!!.toInt()
                    } catch (ignored: NumberFormatException) {
                    }
                }
            } catch (ignored: Exception) {
            }
            return TaskFinder.NOT_FOUND
        }
    }
}
